### PR TITLE
Added Trunk APIs and tests

### PR DESCRIPTION
**Pull request by kashyapjha**

### CODE REVIEW

File: core-test/src/main/java/org/openstack4j/api/network/TrunkTests.java [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+    private static final String JSON_LIST_SUBPORTS_RESPONSE = "/network/listSubPortsResponse.json";
+    private static final String JSON_LIST_TRUNKS_RESPONSE = "/network/listTrunksResponse.json";
+    private static final String JSON_PORT_CREATE_RESPONSE = "/network/portCreateResponse.json";
+
+    @Override
+    protected Service service() {
+        return Service.NETWORK;
+    }
+
+    @Test(enabled = true)
```
Comment (olivergondza) [inline core-test/src/main/java/org/openstack4j/api/network/TrunkTests.java]: `enabled = true` is the default. Please avoid explicitly specifying that.

File: core-test/src/main/java/org/openstack4j/api/network/TrunkTests.java [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+    public void listTrunks() throws Exception {
+        respondWith(JSON_LIST_TRUNKS_RESPONSE);
+
+        String trunk1Id = "cf15956d-4391-4ebf-a9cb-0f7e27b24073";
+        String trunk2Id = "f98559e9-8e92-4100-96ac-a805e0340abd";
+        List<String> trunkIds = new ArrayList<>();
+        assertNotNull(trunkIds);
```
Comment (olivergondza) [inline core-test/src/main/java/org/openstack4j/api/network/TrunkTests.java]: I do not think this needs to be tested.

File: core/src/main/java/org/openstack4j/openstack/networking/domain/NeutronPort.java [hunk 3]
```diff
@@ -232,6 +241,30 @@ public String getvNicType() {
 		return profile;
 	}
 
+	/**
+	 * @return String - the parent ID of the port
```
Comment (olivergondza) [inline core/src/main/java/org/openstack4j/openstack/networking/domain/NeutronPort.java]: There is no need to repeat the return type.

File: core/src/main/java/org/openstack4j/openstack/provider/DefaultAPIProvider.java [hunk 4]
```diff
@@ -171,6 +171,8 @@
 import org.openstack4j.api.trove.InstanceFlavorService;
 import org.openstack4j.api.trove.InstanceService;
 import org.openstack4j.api.trove.TroveService;
+import org.openstack4j.api.networking.TrunkService;
+
```
Comment (olivergondza) [inline core/src/main/java/org/openstack4j/openstack/provider/DefaultAPIProvider.java]: These are unused. please remove

File: core/src/main/java/org/openstack4j/model/network/SubPort.java [hunk 5]
```diff
@@ -0,0 +1,28 @@
+package org.openstack4j.model.network;
+
+import org.openstack4j.common.Buildable;
+import org.openstack4j.model.common.Resource;
+import org.openstack4j.model.network.builder.SubPortBuilder;
+
+/**
+ * A network subport used ONLY in trunks
```
Comment (olivergondza) [inline core/src/main/java/org/openstack4j/model/network/SubPort.java]: Provided this `SubPort` abstraction is only applicable for `Trunk`s. It should be made apparent by either class name (`TrunkSubPort`?), making it a inner class of `Trunk` or moving it to a dedicated package.

File: core/src/main/java/org/openstack4j/model/network/Trunk.java [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+
+    /**
+     * @return the parent port ID of the trunk
+     */
+    String getParentPort();
+
+    /**
+     * @return a list of subports associated with the trunk
+     */
+    List<NeutronSubPort> getSubPorts();
```
Comment (olivergondza) [inline core/src/main/java/org/openstack4j/model/network/Trunk.java]: The interface method should not be returning an implementation, type but its interface instead.

File: core/src/main/java/org/openstack4j/openstack/provider/DefaultAPIProvider.java (hunk 4 above)
Comment (kashyapjha) [reply core/src/main/java/org/openstack4j/openstack/provider/DefaultAPIProvider.java:175]: I forgot to add the binding here that's why they are unused. I'll add in the next commit.

**Comment by kashyapjha** [comment]
Hi Oliver,
I added the changes you requested, could you please review them?

**Comment by kashyapjha** [comment]
Hi Oliver,
The type change from NeutronTrunkSubPort to TrunkSubPort had much more effect than I imagined, please run the tests once.

**Comment by olivergondza** [comment]
Great, did you have a chance to have a look at https://github.com/openstack4j/openstack4j/pull/52#discussion_r402951748?

**Comment by kashyapjha** [comment]
Oh sorry I totally forgot to reply to that one. So basically I just followed how the existing resources do it to maintain consistency.

**Comment by olivergondza** [comment]
@kashyapjha, you are correct this is being done that way sometimes, but it is generally inconsistent all over the codebase. But there are cases when this is done the simple way[1]. Also the method conventions suggest to use name `update` when `Trunk` is updated by `TrunkService`.

[1] https://github.com/openstack4j/openstack4j/blob/628767592d1c70d2cde731e4d2a29e9881606660/core/src/main/java/org/openstack4j/api/identity/v3/UserService.java#L59

**Comment by kashyapjha** [comment]
Also the updateTrunk method takes in a builder built trunk object to update that's why the id is passed in separately

**Comment by olivergondza** [comment]
@kashyapjha, that is my point exactly. We wrap 7 input values in one argument, but we pass the 8th separately. Is there a reason for the ID not being part of the builder and entity?

**Comment by kashyapjha** [comment]
The ID is used to retrieve the current trunk object from openstack and the trunk object which can have all attributes except ID is passed in as the object to update because we cant/shouldnt change a resource ID

**Comment by olivergondza** [comment]
Yes, I understand that's what you are proposing. Though what I suggest is to change that to be aligned with other implementations providing cleaner API for clients. The ID in the `Trunk` will be used as the resource identifier.

[1] https://github.com/openstack4j/openstack4j/blob/628767592d1c70d2cde731e4d2a29e9881606660/core/src/main/java/org/openstack4j/openstack/identity/v3/internal/UserServiceImpl.java#L83

**Comment by kashyapjha** [comment]
Okay I see what you mean. I'll try to implement that way and let you know if it's possible.

**Comment by kashyapjha** [comment]
Hi Oliver,

I've updated it.

**Comment by olivergondza** [comment]
@kashyapjha, I have no idea why, but travis needed a manual rebuild to report test results here. Unfortunately, there are some failing tests.

**Comment by kashyapjha** [comment]
I'm sorry I got a mess of repos and branches so I made a new pull request. All changes you requested are there and unit tests are passing as well

**Comment by olivergondza** [comment]
Closing in favor of #53.

**Pull request closed without merging**